File: contact_service/src/application/dtos.py
from dataclasses import dataclass
from typing import Any, Self

from src.domain.entity import ContactMessage


@dataclass
class ContactMessageDTO:
    id: str
    category: str
    email: str
    name: str
    message: str
    data: dict[str, Any] | None = None
    user_id: str | None = None

    @classmethod
    def from_domain(cls, contact_message: ContactMessage) -> Self:
        return cls(
            id=str(contact_message.id_.value),
            category=contact_message.category.value,
            email=contact_message.email,
            name=contact_message.name,
            message=contact_message.message,
            data=contact_message.data,
            user_id=contact_message.user_id.value if contact_message.user_id else None,
        )

    def to_dict(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "category": self.category,
            "email": self.email,
            "name": self.name,
            "message": self.message,
            "data": self.data,
            "user_id": self.user_id,
        }

File: services/contact-service/src/application/dtos.rs
use std::collections::HashMap;

use crate::domain::entity::ContactMessage;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContactMessageDTO {
    pub id: String,
    pub category: String,
    pub email: String,
    pub name: String,
    pub message: String,
    pub data: Option<HashMap<String, String>>,
    pub received_at: String,
}

impl From<ContactMessage> for ContactMessageDTO {
    fn from(contact_message: ContactMessage) -> Self {
        Self {
            id: contact_message.id.to_string(),
            category: contact_message.category.to_string(),
            email: contact_message.email,
            name: contact_message.name,
            message: contact_message.message,
            data: contact_message.data,
            received_at: contact_message.received_at.to_string(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContactMessageActivityDTO {
    pub count: i64,
    pub date: DateTime<Utc>,
}

impl ContactMessageActivityDTO {
    pub fn new(count: i64, date: DateTime<Utc>) -> Self {
        Self { count, date }
    }
}

impl From<(i64, DateTime<Utc>)> for ContactMessageActivityDTO {
    fn from(value: (i64, DateTime<Utc>)) -> Self {
        Self {
            count: value.0,
            date: value.1,
        }
    }
}
